docs/Practices/hub/hub.md

```diff
@@ -0,0 +1,118 @@
+
+- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Derived Types](./learning.md#derived-types), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Base types representing the shape of domain models should be in the domain directory at the bounded context root. These types can be shared, often for the purpose of derivatives, among slices.
+
+- [UA: Domain Driven Design - Bounded Contexts](./learning.md#bounded-contexts), [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Coupling and Cohesion](./learning.md#coupling-and-cohesion) Domain types, regardless of whether scoped to a slice or scoped to a bounded context, should not depend on code that is not itself either domain code or is a generic utility type (error handling, ts utils, and such). *This is a concept borrowed from Clean/Hexagonal/Onion architecture and key piece in the internal consistency of our domain model if followed. Exceptions to this rule should be very slim or non existent.*
+
+- [UA: Clean Architecture](./learning.md#cleanhexagonalonion-architecture), [UA: Repository pattern](./learning.md#repository-pattern), [UA: Persistence Agnosticism](./learning.md#persistence-agnosticism) Repository layer code is the only code that should depend on types representing the shape of database table/view/query/etc. *The implementation details of data persistence should not be exposed to consumers of a domain model that utilizes said persistence. One of the quickest ways to a building a ball of mud is to do let a database schema (slow and risky to change) be the source of truth for the shape of a domain model across all layers and systems. There are cases where this is ok (constrained problem space, prototypes, etc), but this should be avoided by default.*
+
+- [UA: Slim APIs are Easier to Maintain](./learning.md#slim-apis-are-easier-to-maintain) Don't require more data for a function than is required. *A slim API is less risky to change and easier maintain and reason about.*
```

Comment:
Is it better for an app to make multiple requests to the API for a create call or one call that does multiple actions.

My example is a `createEnrollment` it needs to do many things for this to succeed. Should those be split into multiple API calls for MicroschoolEnrollment, FundingSourceEnrollment and GradeEnrollment? Or is it better that the App does not need to know about each of those?

Reply:
Single call. I'd create a command for `createEnrollment` and then have it call the other commands. If you have to cross system boundaries then of course it would be an HTTP call since we don't have any other mechanisms in place for inter-service communication.

Just avoid reaching into a different services database of course.

*I'm using API here to reference just HTTP APIs. Rather I'm using it generically - it could be the API for a Vertical Slice or module etc. The intended takeaway is if you have to perform an operation that requires properties `x` and `y` from entity `A` that itself has properties `x`, `y`, and `z` - then don't write your endpoint, type signature, or stored procedure etc to take `x`, `y`, AND `z` simply because `z` is a part of entity `A`*